v2/devdoc/log_sink_etw_requirements.md

```diff
@@ -0,0 +1,179 @@
+
+If `message_format` is `NULL`, `log_sink_etw.log_sink_log` shall fail and return.
+
+`log_sink_etw_log` shall register the ETW TraceLogging provider by calling `TraceLoggingRegister` if it was not already done.
+
+`log_sink_etw_log` shall use as provider GUID `DAD29F36-0A48-4DEF-9D50-8EF9036B92B4`.
+
+Note this can (and should) be improved to be configurable later.
+
+`log_sink_etw_log` shall emit a self described event that shall have the name of the event as follows:
+
+  - If `log_level` is `LOG_LEVEL_CRITICAL` the event name shall be `LogCritical`.
+
+  - If `log_level` is `LOG_LEVEL_ERROR` the event name shall be `LogError`.
+
+  - If `log_level` is `LOG_LEVEL_WARNING` the event name shall be `LogWarning`.
+
+  - If `log_level` is `LOG_LEVEL_INFO` the event name shall be `LogInfo`.
+
+  - If `log_level` is `LOG_LEVEL_VERBOSE` the event name shall be `LogVerbose`.
+
+  - Otherwise the event name shall be `Unknown`.
+
+- Logging level:
+
+  - If `log_level` is `LOG_LEVEL_CRITICAL` the ETW logging level shall be `TRACE_LEVEL_CRITICAL`.
+
+  - If `log_level` is `LOG_LEVEL_ERROR` the ETW logging level shall be `TRACE_LEVEL_ERROR`.
+
+  - If `log_level` is `LOG_LEVEL_WARNING` the ETW logging level shall be `TRACE_LEVEL_WARNING`.
+
+  - If `log_level` is `LOG_LEVEL_INFO` the ETW logging level shall be `TRACE_LEVEL_INFO`.
+
+  - If `log_level` is `LOG_LEVEL_VERBOSE` the ETW logging level shall be `TRACE_LEVEL_VERBOSE`.
+
+  - Otherwise the ETW logging level shall be `TRACE_LEVEL_NONE`.
+
+- If `log_context` is `NULL` only the fields `content`, `file`, `func` and `line` shall be added to the ETW event.
+
+- If `log_context` is not `NULL`:
+
+`log_sink_etw.log_sink_log` shall call `log_context_get_property_value_pair_count` to obtain the count of properties that are to be added to the ETW event.
+
+`log_sink_etw.log_sink_log` shall call `log_context_get_property_value_pairs` to obtain the properties that are to be added to the ETW event.
+
+`log_sink_etw.log_sink_log` shall limit the number of properties that are emitted in the ETW event to 64.
+
+`log_sink_etw.log_sink_log` shall compute the metadata size for the self described event metadata as follows:
+
+- Length of the event name (excluding zero terminator) + 1.
+
+- Length of the `content` field name + 1.
+
+- Length of the `file` field name + 1.
+
+- Length of the `func` field name + 1.
+
+- Length of the `line` field name + 1.
+
+- For each property in `log_context`, the length of the property name + 1 and one extra byte for the type of the field.
+
+- For struct properties one extra byte shall be added for the field count.
+
+`log_sink_etw.log_sink_log` shall fill a `SELF_DESCRIBED_EVENT` structure, setting the following fields:
+
+- `_tlgBlobTyp` shall be set to `_TlgBlobEvent4`.
+
+- `_tlgChannel` shall be set to 11.
+
+- `_tlgLevel` shall be set to the appropriate logging level.
+
+- `_tlgOpcode` shall be set to 0.
+
+- `_tlgKeyword` shall be set to 0.
```

Comment:
consider: we can filter ETW collection on level and keyword bitmask, would it be worthwhile to expose this? E.g. set different keywords for different modules like BSDL, BSI, SF replica, etc. Maybe not worth it, just a thought.

Reply:
<html>
<body>
<!--StartFragment-->

ID | Work Item Type | Title | State | Assigned To | Effort | Remaining Work | Original Estimate | Value Area | Iteration Path | Tags
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
17751596 | Task | Add keywork options to etw provider | To Do |   |   |   |   |   | One\Zinc\CY23Q1\Weekly\Wk27 |  

<!--EndFragment-->
</body>
</html>ID	Work Item Type	Title	State	Assigned To	Effort	Remaining Work	Original Estimate	Value Area	Iteration Path	Tags
[17751596](https://msazure.visualstudio.com/One/_workitems/edit/17751596)	Task	Add keywork options to etw provider	To Do	 	 	 	 	 	One\Zinc\CY23Q1\Weekly\Wk27